packages/ra-core/src/controller/useList.ts

```diff
@@ -0,0 +1,233 @@
+import { useCallback, useEffect } from 'react';
+// @ts-ignore
+import { useSelector, useDispatch } from 'react-redux';
+import { parse, stringify } from 'query-string';
+import { push } from 'connected-react-router';
+import pickBy from 'lodash/pickBy';
+
+import queryReducer, {
+    SET_SORT,
+    SET_PAGE,
+    SET_PER_PAGE,
+    SORT_ASC,
+} from '../reducer/admin/resource/list/queryReducer';
+import { crudGetList } from '../actions/dataActions';
+import { changeListParams, ListParams } from '../actions/listActions';
+import { Sort, ReduxState, Identifier, RecordMap } from '../types';
+import { Location } from 'history';
+
+interface Props {
+    filter?: object;
+    filterDefaultValues?: object;
+    perPage?: number;
+    sort?: Sort;
+    location: Location;
+    resource: string;
+}
+
+interface Query extends ListParams {
+    data: RecordMap;
+    ids: Identifier[];
+    total: number;
+    requestSignature: any[];
+}
+
+interface Actions {
+    changeParams: (action: any) => void;
+    setPage: (page: number) => void;
+    setPerPage: (pageSize: number) => void;
+    setSort: (sort: Sort) => void;
+}
+
+const useList = ({
+    resource,
+    location,
+    filterDefaultValues,
+    sort = {
+        field: 'id',
+        order: SORT_ASC,
+    },
+    perPage = 10,
+    filter,
+}: Props): [Query, Actions] => {
+    const dispatch = useDispatch();
+
+    const { params, ids, total } = useSelector(
+        (reduxState: ReduxState) => reduxState.admin.resources[resource].list,
+        [resource]
+    );
+
+    const data = useSelector(
+        (reduxState: ReduxState) => reduxState.admin.resources[resource].data,
+        [resource]
+    );
+
+    const query = getQuery({
+        location,
+        params,
+        filterDefaultValues,
+        sort,
+        perPage,
+    });
+
+    const requestSignature = [
+        resource,
+        JSON.stringify(query),
+        JSON.stringify(location),
+    ];
+
+    const changeParams = useCallback(action => {
+        const newParams = queryReducer(query, action);
+        dispatch(
+            push({
+                ...location,
```

Comment:
I believe you don't have to pass the current location - the `history` package uses the current location by default. That way, if you add the `resource` to the function signature, you can pass an empty array as dependency arguments to `useCallback` and never worry about cache anymore.